DOC: optimize.least_squares: update `curve_fit` docstring and `ValueError` message

m-maggi opened the pull request:
#### Reference issue
Closes #20155.

#### Additional Information
Since I was editing scipy/optimize/_lsq/least_squares.py I also refactored two ".format" strings to f-strings.

m-maggi committed:
DOC optimize.curve_fit docstring

See #20155.

[docs only]

m-maggi committed:
MAINT Refactor ValueError message in optimize.least_squares

See #20155.

[skip ci]

m-maggi committed:
STY Use f-strings instead of ".format" in optimize.least_squares.

[lint only]

j-bowhay commented on scipy/optimize/_minpack_py.py:
```diff
@@ -620,7 +620,7 @@ def curve_fit(f, xdata, ydata, p0=None, sigma=None, absolute_sigma=False,
         initial values will all be 1 (if the number of parameters for the
         function can be determined using introspection, otherwise a
-        ValueError is raised).
+        ValueError is raised). Note that `x0` may be used interchangeably with `p0`.
```
With the change in error message is this statement still needed?

m-maggi replied:
I would remove it. To be on the safe side I committed both a change in the ValueError message and in the documentation. Shall we wait for other opinions?

melissawm replied:
+1 for removing, I think the error message is good now.

mdhaber replied:
```suggestion
        ValueError is raised).
```

mdhaber commented on scipy/optimize/_lsq/least_squares.py:
```diff
@@ -777,8 +777,8 @@ def least_squares(
     if loss not in IMPLEMENTED_LOSSES and not callable(loss):
-        raise ValueError("`loss` must be one of {} or a callable."
-                         .format(IMPLEMENTED_LOSSES.keys()))
+        raise ValueError(f"`loss` must be one of {IMPLEMENTED_LOSSES.keys()} "
+                           "or a callable.")
```
```suggestion
        raise ValueError(f"`loss` must be one of {IMPLEMENTED_LOSSES.keys()} "
                         "or a callable.")
```

mdhaber reviewed (changes requested):
I resolved the merge conflicts and removed the documentation that is no longer needed. (It actually can be read to mean that the user can pass `x0` as a keyword argument interchangeably with `p0`, which would not be correct.) I also added a test that the error message is produced as intended.

Please either [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) or merge m-maggi/scipy#1. If I don't hear back within a week or so, I'll go ahead and open a new PR to fix this.

m-maggi commented:
> I resolved the merge conflicts and removed the documentation that is no longer needed. (It actually can be read to mean that the user can pass `x0` as a keyword argument interchangeably with `p0`, which would not be correct.) I also added a test that the error message is produced as intended.
> 
> Please either [allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) or merge [m-maggi#1](https://github.com/m-maggi/scipy/pull/1). If I don't hear back within a week or so, I'll go ahead and open a new PR to fix this.

Thanks for commenting, I see in this PR that the option "allow edits from maintainers" is ticked (see screenshot below). Should this be ticked in some other place?

![Screenshot 2024-07-04 at 16 43 02](https://github.com/scipy/scipy/assets/124086916/df1c2473-99cd-42e2-8263-13c29588eaf9)

m-maggi committed:
DOC Update scipy/optimize/_minpack_py.py

[Docs only]

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

mdhaber commented:
Thanks @m-maggi. I had written in https://github.com/m-maggi/scipy/pull/1#issuecomment-2212292626 that you'd need to remove the protections from the branch or merge my PR, but I went ahead and opened gh-21147 to avoid further delay. You will still be credited as an author.
